chore: encryption-context changes (#300)
awslabs/aws-encryption-sdk-specification: framework/branch-key-store.md

    - “create-time” (S): `timestamp`
    - "kms-arn" (S): configured KMS Key
    - “hierarchy-version” (N): either `1` or `2`, depending on the `hierarchy-version`

>     - Every key-value pair in the [encryption context](./structures.md#encryption-context) provided by users and is associated with the branch key

Reviewer: This is the creation.
So I think you want to reference "The original input `encryption-context`"

>     - Every key-value pair in the [encryption context](./structures.md#encryption-context) provided by users and is associated with the branch key

Author: ```suggestion
    - Every key-value pair in the `encryption-context` and is associated with the branch key
```

Reviewer: I think you want to reference the prefix. I would way something like all the additional pairs of the branch key context.

    - “create-time” (S): `timestamp`
    - "kms-arn" (S): configured KMS Key
    - “hierarchy-version” (N): either `1` or `2`, depending on the `hierarchy-version`

>     - Every key-value pair in the [encryption context](./structures.md#encryption-context) provided by users and is associated with the branch key

Reviewer: This is the creation.
So I think you want to reference "The original input `encryption-context`"

    - “enc” (B): the wrapped DECRYPT_ONLY Branch Key `CiphertextBlob` from the KMS operation
    - “create-time” (S): `timestamp`
    - "kms-arn" (S): configured KMS Key
    - “hierarchy-version” (N): either `1` or `2`, depending on the `hierarchy-version`

>     - Every key-value pair in the [encryption context](./structures.md#encryption-context) provided by users and is associated with the branch key

Reviewer: This is version key, so it is slightly different.

What is the "encryption context provided by users" at this point? I think we want to be careful at this point. Can we just refer to the branch key context here? And say "Just write the whole context like this"? Because that will already have anything prepended correct?

>     - Every key-value pair in the [encryption context](./structures.md#encryption-context) provided by users and is associated with the branch key

Reviewer: There is some complexity here

>     - Every key-value pair in the [encryption context](./structures.md#encryption-context) provided by users and is associated with the branch key
>       MUST be added with an Attribute Name of `aws-crypto-ec:` + the Key and Attribute Value (S) of the value.

Author: ```suggestion
    - Every other key-value pair in the [branch key context](./structures.md#branch-key-context).
```

    - “enc” (B): wrapped ACTIVE Branch Key `CiphertextBlob` from the KMS operation
    - “create-time” (S): `timestamp`
    - "kms-arn" (S): configured KMS Key
    - “hierarchy-version” (N): either `1` or `2`, depending on the `hierarchy-version`

>     - Every key-value pair in the [encryption context](./structures.md#encryption-context) provided by users and is associated with the branch key

Reviewer: same

>     - Every key-value pair in the [encryption context](./structures.md#encryption-context) provided by users and is associated with the branch key

Author: ```suggestion
    - Every key-value pair in the [Branch Key Context](./structures.md#branch-key-context) that is prepended with `aws-crypto-ec:`
```

Author: We decided on a different suggestion.

Every attribute except `enc` on the AWS DDB response item MUST be converted to a set of key value pairs (a Map)
which is the [branch key context](#branch-key-context).
Every key in the constructed branch key context except tableName MUST exist as a string attribute in the AWS DDB response item.

> The [AWS KMS Encryption Context](#aws-kms-encryption-context) MUST be built from branch key context by extracting keys with prefix `aws-crypto-ec:` and then dropping the prefix.

Reviewer: I think you want a very similar statement that describes how to create the branch key context on Create.

The Beacon key branch key context MUST includes a key `type` and the value MUST be `"beacon:ACTIVE"`.
The Beacon key branch key context MUST NOT have a `version` key.

## AWS KMS Encryption Context

If the `hierarchy-version` is v1, AWS KMS encryption context MUST be same as [branch key context](#branch-key-context).

> If the `hierarchy-version` is v2, AWS KMS encryption context MUST be the [encryption context](./structures.md#encryption-context) sent by users without any transformation. This requirement applies regardless of the Branch Key Item's `type`.

Reviewer: You have this nice section here: #encryption-context-from-authenticated-branch-key-context. It think that this better describes the relationship than "send by the users".

> If the `hierarchy-version` is v2, AWS KMS encryption context MUST be the [encryption context](./structures.md#encryption-context) sent by users without any transformation. This requirement applies regardless of the Branch Key Item's `type`.

Author: I like "sent by users" because for a first time reader its much easier to read.

Author: This is likely what you meant
```suggestion
If the `hierarchy-version` is v2, AWS KMS encryption context MUST be the [encryption context from authenticated branch key context](#encryption-context-from-authenticated-branch-key-context) without any transformation. This requirement applies regardless of the Branch Key Item's `type`.
```

Every attribute except for `enc` on the AWS DDB response item
MUST be authenticated in the decryption of `enc`

Every key in the constructed [branch key context](#branch-key-context)
except `tableName`

> MUST exist as a string attribute in the AWS DDB response item.

Author: Resolved here: [`3dadd20`](https://github.com/awslabs/aws-encryption-sdk-specification/pull/300/commits/3dadd20eecc2f42ea5ae31abde4382c95fdc244f#diff-c6735a6d874d8b9c52040541b81908877a6b9965b51d5b443c35af9895ec8676R797-R799)

  If the KMS Configuration is MRDiscovery, `KeyId` MUST be the `kms-arn` attribute value of the AWS DDB response item, with the region replaced by the configured region.
  Otherwise, it MUST BE the Keystore's configured KMS Key.
- `CiphertextBlob` MUST be the `enc` attribute value on the AWS DDB response item

> - `EncryptionContext` MUST be the [AWS KMS Encryption Context](#aws-kms-encryption-context) constructed above

Reviewer: I don't think you construct that above? But you have the branch key context, and you can use #encryption-context-from-authenticated-branch-key-context

> - `EncryptionContext` MUST be the [AWS KMS Encryption Context](#aws-kms-encryption-context) constructed above

Author: ```suggestion
- `EncryptionContext` MUST be the [encryption context from authenticated branch key context](#encryption-context-from-authenticated-branch-key-context) constructed above
```

Reviewer: I don't think that a section above counts as "constructed above" :)
```suggestion
- `EncryptionContext` MUST be the [encryption context from authenticated branch key context](#encryption-context-from-authenticated-branch-key-context)
```

1. `hierarchy-version`: Version of the hierarchical keyring;
   represented as [AWS DDB Number](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.NamingRulesDataTypes.html#HowItWorks.DataTypes)

A branch key record MAY include [encryption context](./structures.md#encryption-context) key-value pairs.

> These attributes MUST be always prefixed with `aws-crypto-ec:` regardless of the item's `hierarchy-version`.

Reviewer 2: ```suggestion
These attributes MUST be prefixed with `aws-crypto-ec:` regardless of the item's `hierarchy-version`.
```

### Branch Key Materials From Authenticated Branch key Context

The `type` attribute MUST either be equal to `"branch:ACTIVE"` or start with `"branch:version:"`.

If the `type` attribute is equal to `"branch:ACTIVE"`
then the authenticated branch key context MUST have a `version` attribute
and the version string is this value.
If the `type` attribute start with `"branch:version:"` then the version string MUST be equal to this value.

To construct [branch key materials](./structures.md#branch-key-materials) from authenticated branch key context as follows:

- [Branch Key](./structures.md#branch-key) MUST be the [decrypted branch key material](#aws-kms-branch-key-decryption)
- [Branch Key Id](./structures.md#branch-key-id) MUST be the `branch-key-id`
- [Branch Key Version](./structures.md#branch-key-version)
  The version string MUST start with `branch:version:`.
  The remaining string encoded as UTF8 bytes MUST be the Branch Key version.
- [Encryption Context](./structures.md#encryption-context-3) MUST be constructed by
  [Encryption Context From Authenticated Branch Key Context](#encryption-context-from-authenticated-branch-key-context)

> ### Encryption Context From Authenticated Branch Key Context

Reviewer: I think that this is a really good section. I think you can use this everywhere.

The [encryption context](./structures.md#encryption-context) is stored as map of UTF8 Encoded bytes.

For every key in the [encryption context](./structures.md#encryption-context)
the string `aws-crypto-ec:` + the UTF8 decode of this key
MUST exist as a key in the authenticated encryption context.
Also, the value in the [encryption context](./structures.md#encryption-context) for this key
MUST equal the value in the authenticated encryption context

> for the constructed key.

Reviewer 2: Suggestion: "constructed" is vague; prefixed is more explicit.
```suggestion
for the prefixed key.
```
